src/main/java/io/pravega/perf/PerfStats.java

```diff
@@ -94,138 +103,245 @@ public void print(Instant time) {
+        this.csvFile = csvFile;
+        this.queue = new ConcurrentLinkedQueue<TimeStamp>();
+        this.executor = new ForkJoinPool(1);
+    }
+
+    /**
+     * private class for start and end time.
+     */
+    private class ProcessQueue implements Runnable {
+        public void run() {
+            TimeStamp t;
+            while (true) {
+                try {
+                    t = queue.poll();
+                    if (t != null) {
+                        record(t.start, t.end, t.bytes);
+                    } else {
+                        LockSupport.parkNanos(100);
+                    }
+                    print();
+                } catch (IOException ex) {
+                    throw new RuntimeException(ex);
+                }
+            }
         }
-        return values;
+    }
+
+
+    private void record(long startTime, long endTime, int bytes) {
+        final int latency = (int) (endTime - startTime);
+        window.record(bytes, latency);
+        latencyRecorder.record(startTime, bytes, latency);
     }
 
     /**
      * print the performance statistics of current time window.
      */
-    public synchronized void print() throws IOException {
-
-        final Instant time = Instant.now();
+    private void print() throws IOException {
+        final long time = System.currentTimeMillis();
         if (window.windowTimeMS(time) >= windowInterval) {
             window.print(time);
-            this.window = new timeWindow();
-            if (printer != null) {
-                printer.flush();
-            }
+            this.window = new TimeWindow(time);
         }
     }
 
     /**
-     * print the final performance statistics.
+     * start the performance statistics.
+     *
+     * @param startTime start time time
      */
-    public synchronized void printTotal(Instant endTime) throws IOException {
-
-        final double elapsed = Duration.between(start, endTime).toMillis() / 1000.0;
-        final double recsPerSec = count / elapsed;
-        final double mbPerSec = (this.bytes / (1024.0 * 1024.0)) / elapsed;
-
-        long[] percs = percentiles(this.latencies, index, 0.5, 0.95, 0.99, 0.999);
-        System.out.printf(
-                "%d records %s, %.3f records/sec, %d bytes record size, %.2f MB/sec, %.1f ms avg latency, %.1f ms max latency, %d ms 50th, %d ms 95th, %d ms 99th, %d ms 99.9th.\n",
-                count, action, recsPerSec, messageSize, mbPerSec, totalLatency / ((double) count), (double) maxLatency,
-                percs[0], percs[1], percs[2], percs[3]);
-        if (printer != null) {
-            printer.close();
-        }
+    public synchronized void start(long startTime) throws IOException {
+        this.latencyRecorder = csvFile == null ? new LatencyWriter(action, messageSize, startTime) :
+                new CSVLatencyWriter(action, messageSize, startTime, csvFile);
+        this.window = new TimeWindow(startTime);
+        executor.execute(new ProcessQueue());
```

Comment:
Oops this process queue is going to access window, but window was not passed through the queue or under any lock, so line 318 and 316 above might not have run yet from that thread's point of view.

Reply:
ok, now the constructor of the ProcessQueue creates the timewindow and latency recorder